x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -459,6 +460,14 @@ public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
     public List<Setting<?>> getSettings() {
         List<Setting<?>> settings = super.getSettings();
         settings.add(SourceOnlySnapshotRepository.SOURCE_ONLY);
+
+        // Don't register the license setting if there is an alternate implementation loaded as an extension.
+        // this relies on the order in which methods are called - loadExtensions, (this method) getSettings, then createComponents
+        if (getSharedLicenseService() == null) {
+            settings.add(LicenseSettings.SELF_GENERATED_LICENSE_TYPE);
+            settings.add(LicenseSettings.ALLOWED_LICENSE_TYPES_SETTING);
+        }
```

Comment:
I am a bit concerned if `createComponents` ever gets called before this method, we would have a bug that these license settings are not registered. Also when the settings are not registered, we get a generic error message like:

> unknown setting [xxx] please check that any required plugins are installed, or check the breaking changes documentation for removed settings

It is not particularly correct in this case. 

Instead of removing them, I wonder whether we could validate them in `createComponents`, e.g. something like
```java
if (licenseService == null) {
   // create the cluster state license service as is
} else {
   // new code
   if (LicenseSettings.SELF_GENERATED_LICENSE_TYPE.exists(settings)) {
       throw new IllegalArgumentException("setting is not supported when a license plugin is in use")
   }
}
```

This would also make it eaiser to write a test. 

Or as a bare minimum, we could choose to do nothing. These settings are ignored when the license plugin is in use. So they are harmless. One analogy is when security is disabled, we still register all security related settings which are simply ignored. This behaviour is for usability because user might be experimenting with security and we don't want to force a large set of configuration changes every time security is disabled.

Reply:
On a second thought, many existing tests use the `SELF_GENERATED_LICENSE_TYPE` setting for a trial license. So if there is ever a change in order of method calls that results into settings not registered, these tests will all fail. Effectively, we already have tests that guarantee the settings are registered when they are needed. This resolves my major concern here. Still happy to discuss other aspects like error message. But they are not worth to block this PR.

Reply:
The order in which these methods are called is an implicit guarantee ... if it ever change it would break ALOT of things beyond this and would break dozens of tests for this license. I don't like that we have to rely on implicit ordering for initialization but it is better than doing nothing and registering them only to throw an error if you set it seems unnecessary. The user will never see that error message and any errors that may appear are due to either a bug in source code or misconfiguration by an Elastic employee working on an internal only project.  Allowing these to exist , be set, and do nothing is a form a lenience that we typically try to avoid. 

> One analogy is when security is disabled, we still register all security related settings which are simply ignored.

I was not aware of this ... and IMO we should fix that ASAP. To my knowledge we don't do that anywhere else and sometimes go to near extreme measure to ensure that groups of settings all work together and don't allow misconfiguration.  We should not have settings you can set but that do nothing. It can lead to all kinds of misunderstanding, misconfiguration and frustration which can be especially dangerous with security configuration.

Reply:
collision of comments :)  ... thanks Yang for the review.  Let's follow up outside of this PR on the settings you can set but don't do anything.